src/build/render/index.js

```diff
@@ -0,0 +1,78 @@
+const path = require("path");
+const fs = require("fs").promises;
+const R = require("ramda");
+
+const {wrapper} = require("./components");
+const features = require("./features");
+
+async function renderPage(ctx) {
+  const {page, lang} = ctx;
+
+  /* Pass the render context through each feature, gathering their desired
+   * additions to the page and search index. We use the `combineResults` helper
+   * to grab a result key from all features if present, then merge using a
+   * provided pipeline of functions.
+   */
+  const featureResults = await Promise.all(features.map(feature => feature(ctx)));
+  const combineResults = (key, ...pipe) => R.pipe(
+    R.map(R.prop(key)),
+    R.filter(val => !R.isNil(val)),
+    ...pipe
+  )(featureResults);
+
+  const metaboxProps = {
+    metaTitle: combineResults("metaTitle", R.last),
+    img: combineResults("img", R.last),
+    imgCaption: combineResults("imgCaption", R.last),
+    metaClass: combineResults("metaClass", R.last),
+    metaSections: combineResults("metaSections", R.flatten),
+  };
+
+  //represents the full page HTML, ready to write to a file
+  const htmlDoc = wrapper(
+    ctx,
+    combineResults("headings", R.flatten),
+    combineResults("thanks", R.reduce(R.mergeWith(R.concat), {})),
+    metaboxProps,
+    combineResults("html")
+  );
+  //used to add the page to the search index
+  const searchDoc = page.noSearch ? null : {
+    lang,
+    path: ctx.page.localizedPaths[ctx.lang],
+    title: ctx.page.title[ctx.lang],
+    text: combineResults("searchText", R.join(" ")),
+    keywords: combineResults("keywords", R.flatten, R.join(" "))
+  };
```

Comment:
It's here where all the requested additions from each render feature are merged together and used to construct the resulting page and search document.